File: easy_1.cpp
#include <iostream>
#include <set>
using namespace std;

int HuffmanCoding(vector<int> a) {
    int ans = 0;
    priority_queue<int, vector<int>, greater<int> > freqs;

    for (int freq : a) {
        freqs.push(freq);
    }

    while (freqs.size() > 1) {
        int first = freqs.top(); freqs.pop();
        int second = freqs.top(); freqs.pop();
        ans += first + second;
        freqs.push(first + second);
    }

    return ans;
}

int main() {
    string text;
    cin >> text;

    set<char> uniqueElements (text.begin(), text.end());
    string s(uniqueElements.begin(), uniqueElements.end());
    vector<int> count_letters;

    for (char c : s) {
        count_letters.push_back(count(text.begin(), text.end(), c));
    }
    cout << HuffmanCoding(count_letters) << '\n';

    return 0;
}

File: easy_1.py
import heapq

def huffman_coding(text):
    # Подсчет частоты символов
    frequencies = {}
    for char in text:
        frequencies[char] = frequencies.get(char, 0) + 1

    # Преобразование частот в список
    freq_list = list(frequencies.values())
    heapq.heapify(freq_list)

    total_cost = 0
    while len(freq_list) > 1:
        first = heapq.heappop(freq_list)
        second = heapq.heappop(freq_list)
        total_cost += first + second
        heapq.heappush(freq_list, first + second)

    return total_cost;

print(huffman_coding("2h2.+.o2.=.2h2o"))
